src/kokkos/KokkosCore/HistoContainer.h

```diff
@@ -0,0 +1,339 @@
+#ifndef HeterogeneousCore_CUDAUtilities_interface_HistoContainer_h
+#define HeterogeneousCore_CUDAUtilities_interface_HistoContainer_h
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+
+#include "KokkosCore/AtomicPairCounter.h"
+
+using team_policy = Kokkos::TeamPolicy<KokkosExecSpace>;
+using member_type = Kokkos::TeamPolicy<KokkosExecSpace>::member_type;
+
+namespace cms {
+  namespace kokkos {
+    template <typename Histo, typename T>
+    KOKKOS_FUNCTION void countFromVector(Kokkos::View<Histo*,KokkosExecSpace> h,
+                                         const uint32_t nh,
+                                         Kokkos::View<T const*,KokkosExecSpace> v,
+                                         Kokkos::View<uint32_t const*,KokkosExecSpace> offsets,
+                                         const member_type& teamMember) {
+      uint32_t first = teamMember.league_rank() * teamMember.team_size() + teamMember.team_rank();
+      uint32_t total_threads = teamMember.league_size() * teamMember.team_size();
+      for (uint32_t i = first, nt = offsets(nh); i < nt; i += total_threads) {
+        uint32_t index = 0;
+        for(uint32_t j = 0;j<=nh;++j){
+          if(offsets(j) > i){
+             index = j;
+             break;
+          }
+        }
+        assert(offsets(index) > 0);
+        int32_t ih = &offsets(index) - &offsets(0) - 1;
+        assert(ih >= 0);
+        assert(ih < int(nh));
+        h(0).count(v(i), ih);
+      }
+    }
+
+
+    template <typename Histo, typename T>
+    KOKKOS_FUNCTION void fillFromVector(Kokkos::View<Histo*,KokkosExecSpace> h,
+                                        uint32_t nh,
+                                        Kokkos::View<T const*,KokkosExecSpace> v,
+                                        Kokkos::View<uint32_t const*,KokkosExecSpace> offsets,
+                                        const member_type& teamMember) {
+      int first = teamMember.league_rank() * teamMember.team_size() + teamMember.team_rank();
+      int total_threads = teamMember.league_size() * teamMember.team_size();
+
+      for (uint32_t i = first, nt = offsets(nh); i < nt; i += total_threads) {
+        uint32_t index = 0;
+        for(uint32_t j = 0;j<=nh;++j){
+          if(offsets(j) > i){
+             index = j;
+             break;
+          }
+        }
+        assert(offsets(index) > 0);
+        int32_t ih = &offsets(index) - &offsets(0) - 1;
+        assert(ih >= 0);
+        assert(ih < int(nh));
+        h(0).fill(v(i), i, ih);
+      }
+    }
+
+    template <typename Histo>
+    inline void launchZero(Kokkos::View<Histo*,KokkosExecSpace> h) {
+      Kokkos::parallel_for(Histo::totbins(),KOKKOS_LAMBDA(const size_t i) {
```

Comment:
I think this would be
```c++
Kokkos::parallel_for(Kokkos::RangePolicy<execSpace>(0,Histo::totbins()), KOKKOS_LAMBDA(const size_t& i) {
```

Reply:
It can't be `Kokkos::RangePolicy<execSpace>(...)` because `execSpace` is a variable and a type. The point is to pass the `execSpace` object to the `RangePolicy` constructor.

Reply:
Sorry I meant this:
```python
template <typename Histo, typename ExecSpace>
    inline void launchZero(Kokkos::View<Histo*,ExecSpace> h, ExecSpace const& execSpace) {
      Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0,Histo::totbins()), 
        KOKKOS_LAMBDA(const size_t i) {
```
Doesn't this tell the range policy to execute in the same space as the histo view is located? And why are we passing the `execSpace` object to the `RangePolicy`?

If I follow your example which has `RangePolicy<KokkosExecSpace>(execSpace)` doesn't using the `KokkosExecSpace` negate the point?

Reply:
When execution space object is passed to the execution policy constructor, the policy internally will call `fence()` only on that execution space instead of global `Kokkos::fence()`. It follows that passing execution space object always is the only way to allow running a serial-only job of a serial+CUDA build on a machine without a GPU (otherwise `Kokkos::fence()` calls `fence()` also on CUDA space, which fails without a GPU).

Furthermore, passing execution space object is necessary for asynchronous execution (to which we'll hopefully get some day), and crafting everything towards that now saves later effort.

Reply:
OK, updated this to use `ExecSpace` in template and `execSpace` object passed to constructor.